pkg/statdb/statdb.go

```diff
@@ -0,0 +1,227 @@
+func (s *Server) Get(ctx context.Context, getReq *pb.GetRequest) (resp *pb.GetResponse, err error) {
+	s.logger.Debug("entering statdb Get")
+
+	apiKeyBytes := []byte(getReq.ApiKey)
+	err = s.validateAuth(apiKeyBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	dbNode, err := s.DB.Get_Node_By_Id(ctx, dbx.Node_Id(string(getReq.NodeId)))
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, err.Error())
+	}
+
+	nodeStats := &pb.NodeStats{
+		NodeId:            []byte(dbNode.Id),
+		AuditSuccessRatio: dbNode.AuditSuccessRatio,
+		UptimeRatio:       dbNode.UptimeRatio,
+	}
+	return &pb.GetResponse{
+		Stats: nodeStats,
+	}, nil
+}
+
+// Update a single farmer's stats in the db
+func (s *Server) Update(ctx context.Context, updateReq *pb.UpdateRequest) (resp *pb.UpdateResponse, err error) {
+	s.logger.Debug("entering statdb Update")
+
+	apiKeyBytes := []byte(updateReq.ApiKey)
+	err = s.validateAuth(apiKeyBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	node := updateReq.Node
+
+	dbNode, err := s.DB.Get_Node_By_Id(ctx, dbx.Node_Id(string(node.NodeId)))
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, err.Error())
+	}
+
+	auditSuccessCount := dbNode.AuditSuccessCount
+	totalAuditCount := dbNode.TotalAuditCount
+	auditSuccessRatio := dbNode.AuditSuccessRatio
+	uptimeSuccessCount := dbNode.UptimeSuccessCount
+	totalUptimeCount := dbNode.TotalUptimeCount
+	uptimeRatio := dbNode.UptimeRatio
+
+	updateFields := dbx.Node_Update_Fields{}
+
+	if node.UpdateAuditSuccess {
+		auditSuccessCount, totalAuditCount, auditSuccessRatio = updateRatioVars(
+			node.AuditSuccess,
+			auditSuccessCount,
+			totalAuditCount,
+		)
+
+		updateFields.AuditSuccessCount = dbx.Node_AuditSuccessCount(auditSuccessCount)
+		updateFields.TotalAuditCount = dbx.Node_TotalAuditCount(totalAuditCount)
+		updateFields.AuditSuccessRatio = dbx.Node_AuditSuccessRatio(auditSuccessRatio)
+	}
+	if node.UpdateUptime {
+		uptimeSuccessCount, totalUptimeCount, uptimeRatio = updateRatioVars(
+			node.IsUp,
+			uptimeSuccessCount,
+			totalUptimeCount,
+		)
+
+		updateFields.UptimeSuccessCount = dbx.Node_UptimeSuccessCount(uptimeSuccessCount)
+		updateFields.TotalUptimeCount = dbx.Node_TotalUptimeCount(totalUptimeCount)
+		updateFields.UptimeRatio = dbx.Node_UptimeRatio(uptimeRatio)
+	}
+
+	dbNode, err = s.DB.Update_Node_By_Id(ctx, dbx.Node_Id(string(node.NodeId)), updateFields)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, err.Error())
+	}
+
+	nodeStats := &pb.NodeStats{
+		NodeId:            []byte(dbNode.Id),
+		AuditSuccessRatio: dbNode.AuditSuccessRatio,
+		UptimeRatio:       dbNode.UptimeRatio,
+	}
+	return &pb.UpdateResponse{
+		Stats: nodeStats,
+	}, nil
+}
+
+// UpdateBatch for updating  multiple farmers' stats in the db
+func (s *Server) UpdateBatch(ctx context.Context, updateBatchReq *pb.UpdateBatchRequest) (resp *pb.UpdateBatchResponse, err error) {
+	s.logger.Debug("entering statdb UpdateBatch")
+
+	apiKeyBytes := []byte(updateBatchReq.ApiKey)
```

Comment:
do we want to validate the APIKey ?

Reply:
The api key is validated inside the `Update` function, which `UpdateBatch` calls. So it will be validated for each node in the batch, and it is unnecessary to call it here.

Reply:
see it- thanks!